lnwire/accept_channel.go

```diff
@@ -115,7 +116,7 @@ var _ Message = (*AcceptChannel)(nil)
 // protocol version.
 //
 // This is part of the lnwire.Message interface.
-func (a *AcceptChannel) Encode(w io.Writer, pver uint32) error {
+func (a *AcceptChannel) Encode(w *bytes.Buffer, pver uint32) error {
```

Comment:
So the rationale here is that we end up saving a heap escape? I guess in the end, we don't ever really write _directly_ to the wire (where the `io.Writer` could be useful), since we need to encrypt the message payload first using our noise construction.

Reply:
The `Encode` calls `WriteElements`, which saves several heap escapes when using a concrete type. I think this is where the gain comes from, and the incremental gain from this commit compared to the previous one,
```
name                                    old allocs/op  new allocs/op  delta
WriteMessage/Init-4                         6.00 ± 0%      6.00 ± 0%     ~     (all equal)
WriteMessage/Error-4                        4.00 ± 0%      2.00 ± 0%  -50.00%  (p=0.000 n=10+10)
WriteMessage/Ping-4                         4.00 ± 0%      2.00 ± 0%  -50.00%  (p=0.000 n=10+10)
WriteMessage/Pong-4                         2.00 ± 0%      1.00 ± 0%  -50.00%  (p=0.000 n=10+10)
WriteMessage/MsgOpenChannel-4               48.0 ± 0%      32.0 ± 0%  -33.33%  (p=0.000 n=10+10)
WriteMessage/MsgAcceptChannel-4             42.0 ± 0%      29.0 ± 0%  -30.95%  (p=0.000 n=10+10)
WriteMessage/MsgFundingCreated-4            6.00 ± 0%      3.00 ± 0%  -50.00%  (p=0.000 n=10+10)
WriteMessage/MsgFundingSigned-4             4.00 ± 0%      2.00 ± 0%  -50.00%  (p=0.000 n=10+10)
WriteMessage/FundingLocked-4                7.00 ± 0%      5.00 ± 0%  -28.57%  (p=0.000 n=10+10)
WriteMessage/Shutdown-4                     4.00 ± 0%      2.00 ± 0%  -50.00%  (p=0.000 n=10+10)
WriteMessage/ClosingSigned-4                6.00 ± 0%      3.00 ± 0%  -50.00%  (p=0.000 n=10+10)
WriteMessage/UpdateAddHTLC-4                10.0 ± 0%       6.0 ± 0%  -40.00%  (p=0.000 n=10+10)
WriteMessage/UpdateFulfillHTLC-4            5.00 ± 0%      3.00 ± 0%  -40.00%  (p=0.000 n=10+10)
WriteMessage/UpdateFailHTLC-4               5.00 ± 0%      2.00 ± 0%  -60.00%  (p=0.000 n=10+10)
WriteMessage/CommitSig-4                     208 ± 0%       105 ± 0%  -49.52%  (p=0.000 n=10+10)
WriteMessage/RevokeAndAck-4                 8.00 ± 0%      6.00 ± 0%  -25.00%  (p=0.000 n=10+10)
WriteMessage/UpdateFee-4                    4.00 ± 0%      2.00 ± 0%  -50.00%  (p=0.000 n=10+10)
WriteMessage/UpdateFailMalformedHTLC-4      8.00 ± 0%      5.00 ± 0%  -37.50%  (p=0.000 n=10+10)
WriteMessage/ChannelReestablish-4           10.0 ± 0%       6.0 ± 0%  -40.00%  (p=0.000 n=10+10)
WriteMessage/ChannelAnnouncement-4          25.0 ± 0%      14.0 ± 0%  -44.00%  (p=0.000 n=10+10)
WriteMessage/NodeAnnouncement-4             33.0 ± 0%      15.0 ± 0%  -54.55%  (p=0.000 n=10+10)
WriteMessage/ChannelUpdate-4                22.0 ± 0%      10.0 ± 0%  -54.55%  (p=0.000 n=10+10)
WriteMessage/AnnounceSignatures-4           12.0 ± 0%       6.0 ± 0%  -50.00%  (p=0.000 n=10+10)
WriteMessage/QueryShortChanIDs-4           4.01k ± 0%     1.00k ± 0%  -74.92%  (p=0.000 n=10+10)
WriteMessage/ReplyShortChanIDsEnd-4         2.00 ± 0%      1.00 ± 0%  -50.00%  (p=0.000 n=10+10)
WriteMessage/QueryChannelRange-4            5.00 ± 0%      3.00 ± 0%  -40.00%  (p=0.000 n=10+10)
WriteMessage/ReplyChannelRange-4           4.01k ± 0%     1.00k ± 0%  -74.94%  (p=0.000 n=10+10)
WriteMessage/GossipTimestampRange-4         5.00 ± 0%      3.00 ± 0%  -40.00%  (p=0.000 n=10+10)
WriteMessage/QueryShortChanIDs#01-4        4.03k ± 0%     1.04k ± 0%  -74.29%  (p=0.000 n=10+10)
WriteMessage/ReplyChannelRange#01-4        4.03k ± 0%     1.04k ± 0%  -74.31%  (p=0.000 n=10+10)
```

> I guess in the end, we don't ever really write directly to the wire (where the `io.Writer` could be useful)

A bit confused here🧐 If the `io.Writer` works, so does `*bytes.Buffer` right?